### PR TITLE
Google OAuth Implementation

### LINKED ISSUE #49: Implement Google Authentication with Passport
**Description:**
As part of enhancing our application's authentication system, we need to implement Google Authentication using Passport.js and JWT (JSON Web Tokens). This will enable users to log in securely using their Google accounts and manage their sessions using JWT tokens.

**Tasks:**
1. Set up Passport.js and JWT dependencies in the project to handle Google Authentication.
2. Configure Passport.js with the Google strategy to enable authentication via Google accounts.
3. Implement user registration endpoint to create new user accounts securely.
4. Create a Google login endpoint that authenticates users using their Google accounts and generates JWT tokens.
5. Implement a protected endpoint that requires a valid JWT token for access.
6. Set up middleware to authenticate requests using Passport.js and the JWT strategy.
7. Handle logout functionality to invalidate tokens or sessions.
8. Implement password reset functionality for users who have forgotten their password.

**Acceptance Criteria:**
- Users should be able to register and create new accounts securely.
- Users should be able to log in using their Google accounts and receive a JWT token for subsequent authenticated requests.
- The protected endpoint should only be accessible with a valid JWT token.
- JWT tokens should be properly validated, and expired or invalid tokens should be rejected.
- Users should be able to log out and invalidate their session or token.
- The password reset functionality should allow users to reset their password securely.

**Additional Information:**
- Make sure to follow best practices for security, such as properly hashing passwords and securely handling JWT tokens.
- Document the authentication API endpoints and any required input parameters or response formats.
- Ensure that appropriate error handling is implemented for various scenarios, such as invalid credentials or expired tokens.
- Users' personal information obtained from Google should be handled with care and in compliance with privacy regulations.
- Provide clear instructions in the documentation on how to set up and configure Google Authentication for the application.
- Provide clear instructions on how to test the authentication endpoints and verify the functionality.

**Related Dependencies or References:**
- Passport.js: [link to documentation](https://www.passportjs.org/)
- JWT: [link to documentation](https://jwt.io/)
- Google OAuth2 API: [link to documentation](https://developers.google.com/identity/protocols/oauth2)

### PR DESCRIPTION
**Pull request by mayura-andrew**
## Purpose
The purpose of this PR is to fix #49 

## Goals
- To implementing Google OAuth api endpoint

## Approach
 - used passportjs googleoauth strategy for this google authentication.
 
## Screenshots
[Screencast_20240119_215209.webm](https://github.com/sef-global/scholarx-backend/assets/48531182/09065425-dae2-4dab-b93f-a7b988715e7a)

## Checklist
- [] This PR doesn't commit any keys, passwords, tokens, usernames, or other secrets.
- [x] I have read and understood the development best practices guidelines (http://bit.ly/sef-best-practices)
- [x] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [x] I have made corresponding changes to the documentation

## Related PRs
N/A

## Test environment
- Node.js v20.10.0

## Learning
While working on this issue, I learned more about how to implement google OAuth using passportjs.

### CODE REVIEW

File: package.json [hunk 1]
```diff
@@ -34,11 +38,14 @@
...
     "@types/cors": "^2.8.13",
     "@types/express": "^4.17.17",
+    "@types/express-session": "^1.17.10",
     "@types/jest": "^29.5.3",
     "@types/jsonwebtoken": "^9.0.2",
     "@types/node": "^20.1.4",
     "@types/passport": "^1.0.12",
+    "@types/passport-google-oauth20": "^2.0.14",
     "@types/passport-jwt": "^3.0.9",
+    "@types/passport-linkedin-oauth2": "^1.5.6",
```
Comment (anjula-sack) [inline package.json:48]:
```suggestion
```

File: package.json [hunk 2]
```diff
@@ -20,9 +20,13 @@
...
     "dotenv": "^16.3.1",
     "express": "^4.18.2",
-    "jsonwebtoken": "^9.0.1",
+    "express-session": "^1.17.3",
+    "jsonwebtoken": "^9.0.2",
     "passport": "^0.6.0",
+    "passport-google-oauth20": "^2.0.0",
+    "passport-google-oidc": "^0.1.0",
     "passport-jwt": "^4.0.1",
+    "passport-linkedin-oauth2": "^2.0.0",
```
Comment (anjula-sack) [inline package.json:29]:
```suggestion
```

File: src/configs/passport.ts [hunk 3]
```diff
@@ -2,15 +2,59 @@ import passport from 'passport'
...
+        const user = await findOrCreateUser(profile)
+        done(null, user)
+      } catch (err) {
+        done(err as Error)
+      }
+    }
+  )
+)
+
+passport.serializeUser((user: any, done) => {
```
Comment (anjula-sack) [inline src/configs/passport.ts:36]: Let's avoid using any type

```suggestion
passport.serializeUser((user: User, done) => {
```

File: src/configs/passport.ts [hunk 4]
```diff
@@ -2,15 +2,59 @@ import passport from 'passport'
...
+  try {
+    const profileRepository = dataSource.getRepository(Profile)
+    const user = await profileRepository.findOne({ where: { primary_email } })
+    done(null, user)
+  } catch (err) {
+    done(err)
+  }
+})
+
+// -----
```
Comment (anjula-sack) [inline src/configs/passport.ts:50]:
```suggestion
```

File: src/configs/passport.ts [hunk 5]
```diff
@@ -2,15 +2,59 @@ import passport from 'passport'
...
 const cookieExtractor = (req: Request): string => {
   let token = null
   if (req?.cookies) {
     token = req.cookies.jwt
   }
   return token
+  console.log(token)
```
Comment (anjula-sack) [inline src/configs/passport.ts:57]:
```suggestion
```

File: tsconfig.json [hunk 6]
```diff
@@ -1,6 +1,6 @@
 {
   "compilerOptions": {
-    "target": "es6",
+    "target": "es2022",
```
Comment (anjula-sack) [inline tsconfig.json:3]: Why has this changed @mayura-andrew?

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: tsconfig.json (hunk 6 above)
Comment (mayura-andrew) [reply tsconfig.json:3]: for some testing purposes i'll remove this

File: src/configs/passport.ts (hunk 3 above)
Comment (mayura-andrew) [reply src/configs/passport.ts:36]: if i need to add User type first we need to write a User interface right ?

File: .vscode/settings.json [hunk 7]
```diff
@@ -0,0 +1,5 @@
+{
```
Comment (anjula-sack) [inline .vscode/settings.json:1]: Can you remove this file? @mayura-andrew

File: .env.example [hunk 8]
```diff
@@ -5,3 +5,6 @@ DB_PASSWORD=your_db_password
 DB_PORT=your_db_port_number
 SERVER_PORT=your_server_port_number
 JWT_SECRET=your_jwt_secret_key
+GOOGLE_CLIENT_ID='your_google_client_id'
+GOOGLE_CLIENT_SECRET='your_google_client_secret'
+GOOGLE_REDIRECT_URL=http://localhost:3000/api/auth/google/callback
```
Comment (anjula-sack) [inline .env.example:10]: Add the EOL

File: src/configs/passport.ts [hunk 9]
```diff
@@ -2,9 +2,54 @@ import passport from 'passport'
...
+  GOOGLE_CLIENT_ID,
+  GOOGLE_REDIRECT_URL,
+  GOOGLE_CLIENT_SECRET
+} from './envConfig'
 import type { Request } from 'express'
 
+// Google Authencation Strategy  ----
+import { Strategy as GoogleStrategy } from 'passport-google-oauth20'
+import { findOrCreateUser } from '../services/auth.service'
+// Import the 'User' type from the appropriate module
```
Comment (anjula-sack) [inline src/configs/passport.ts:16]:
```suggestion
```

File: src/controllers/auth.controller.ts [hunk 10]
```diff
@@ -6,6 +6,38 @@ import jwt from 'jsonwebtoken'
...
+          return
+        }
+        console.log('login success')
+        res.redirect('/')
+      })
+    }
+  )(req, res, next)
+}
+
+// -----
```
Comment (anjula-sack) [inline src/controllers/auth.controller.ts:40]:
```suggestion
```

File: src/controllers/auth.controller.ts [hunk 11]
```diff
@@ -6,6 +6,38 @@ import jwt from 'jsonwebtoken'
...
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
+export const googleRedirect = async (
+  req: Request,
+  res: Response,
+  next: NextFunction
+) => {
+  passport.authenticate(
+    'google',
+    { failureRedirect: '/login' },
+    (err: any, user: any, info: any) => {
```
Comment (anjula-sack) [inline src/controllers/auth.controller.ts:19]: Let's define the types here

File: src/controllers/auth.controller.ts [hunk 12]
```diff
@@ -6,6 +6,38 @@ import jwt from 'jsonwebtoken'
 import { JWT_SECRET } from '../configs/envConfig'
 import type { ApiResponse } from '../types'
 
+// Google Authencation  Redirect code ----
+// eslint-disable-next-line @typescript-eslint/explicit-function-return-type
```
Comment (anjula-sack) [inline src/controllers/auth.controller.ts:10]:
```suggestion
```

File: src/configs/passport.ts [hunk 13]
```diff
@@ -2,9 +2,54 @@ import passport from 'passport'
...
-import { JWT_SECRET } from './envConfig'
+import {
+  JWT_SECRET,
+  GOOGLE_CLIENT_ID,
+  GOOGLE_REDIRECT_URL,
+  GOOGLE_CLIENT_SECRET
+} from './envConfig'
 import type { Request } from 'express'
 
+// Google Authencation Strategy  ----
```
Comment (anjula-sack) [inline src/configs/passport.ts:13]:
```suggestion
```

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: .vscode/settings.json [hunk 14]
```diff
@@ -0,0 +1,5 @@
...
+    "githubPullRequests.ignoredPullRequestBranches": [
+        "main"
+    ]
+}
```
Comment (YoshithaRathnayake) [inline .vscode/settings.json:5]: > Add the EOL

Add the EOL here also @mayura-andrew

**Review by YoshithaRathnayake — CHANGES_REQUESTED** [review]

File: src/services/auth.service.ts [hunk 15]
```diff
@@ -83,3 +83,27 @@ export const loginUser = async (
     return { statusCode: 500, message: 'Internal server error' }
   }
 }
+
+// findOrCreateUser function for Google OAuth
```
Comment (anjula-sack) [inline src/services/auth.service.ts:87]:
```suggestion
```

File: src/services/auth.service.ts [hunk 16]
```diff
@@ -83,3 +83,27 @@ export const loginUser = async (
...
+export const findOrCreateUser = async (profile: any): Promise<Profile> => {
```
Comment (anjula-sack) [inline src/services/auth.service.ts:88]: Add the type here

File: .env.example (hunk 8 above)
Comment (anjula-sack) [inline .env.example:10]: The port is depending on SERVER_PORT.

**Review by anjula-sack — CHANGES_REQUESTED** [review]

File: src/services/auth.service.ts (hunk 16 above)
Comment (mayura-andrew) [reply src/services/auth.service.ts:88]: when i add type for async (profile: any) i'm getting type error in passport.ts

File: src/configs/passport.ts [hunk 17]
```diff
@@ -2,9 +2,57 @@ import passport from 'passport'
...
+  new GoogleStrategy(
+    {
+      clientID: GOOGLE_CLIENT_ID,
+      clientSecret: GOOGLE_CLIENT_SECRET,
+      callbackURL: GOOGLE_REDIRECT_URL,
+      scope: ['profile', 'email'],
+      passReqToCallback: true // Add this line
+    },
+    async function (
+      req: Request, // Add this line
```
Comment (anjula-sack) [inline src/configs/passport.ts:27]:
```suggestion
      req: Request,
```
 Remove the comments in this section as well @mayura-andrew

**Review by anjula-sack — CHANGES_REQUESTED** [review]

**Review by anjula-sack — APPROVED** [review]
Good job! @mayura-andrew

**Pull request merged**